dev/bots/run_command.dart

```diff
@@ -2,92 +2,112 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:async';
 import 'dart:convert';
 import 'dart:core' hide print;
-import 'dart:io' hide exit;
+import 'dart:io' as io;
 
 import 'package:path/path.dart' as path;
 
 import 'utils.dart';
 
-// TODO(ianh): These two functions should be refactored into something that avoids all this code duplication.
-
+/// Runs the `executable` and returns standard output as a stream of lines.
+///
+/// The returned stream reaches its end immediately after the command exits.
+///
+/// If `expectNonZeroExit` is false and the process exits with a non-zero exit
+/// code fails the test immediately by exiting the test process with exit code
+/// 1.
 Stream<String> runAndGetStdout(String executable, List<String> arguments, {
   String workingDirectory,
   Map<String, String> environment,
   bool expectNonZeroExit = false,
-  int expectedExitCode,
-  String failureMessage,
-  bool skip = false,
 }) async* {
-  final String commandDescription = '${path.relative(executable, from: workingDirectory)} ${arguments.join(' ')}';
-  final String relativeWorkingDir = path.relative(workingDirectory);
-  if (skip) {
-    printProgress('SKIPPING', relativeWorkingDir, commandDescription);
-    return;
-  }
-  printProgress('RUNNING', relativeWorkingDir, commandDescription);
-
-  final Stopwatch time = Stopwatch()..start();
-  final Process process = await Process.start(executable, arguments,
+  final StreamController<String> output = StreamController<String>();
+  final Future<CommandResult> command = runCommand(
+    executable,
+    arguments,
     workingDirectory: workingDirectory,
     environment: environment,
+    expectNonZeroExit: expectNonZeroExit,
+    // Capture the output so it's not printed to the console by default.
+    outputMode: OutputMode.capture,
+    outputListener: (String line, io.Process process) {
+      output.add(line);
+    },
   );
 
-  stderr.addStream(process.stderr);
-  final Stream<String> lines = process.stdout.transform(utf8.decoder).transform(const LineSplitter());
-  yield* lines;
+  yield* output.stream;
```

Comment:
What does this actually do?

Reply:
I'm trying to mentally model this - we run the stream until it gets closed, then await the result?

But when does the stream get closed?

Reply:
I imagine the stream gets closed when the process exit. I just don't know which one resolves first the stream or the exit code :man_shrugging:

Reply:
But you created the output stream controller above, right?

Reply:
Are you implying that I need to explicitly close the controller? Hmm :thinking:

Reply:
other wise how does the stream end 😓

Reply:
Yep, you were right. Done.

Reply:
For some reason I thought GC finalizes it... although now I realize that wouldn't work in JS until very recently.